plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/reload/SpringBeanReload.java

```diff
@@ -0,0 +1,608 @@
+package org.hotswap.agent.plugin.spring.reload;
+
+import org.hotswap.agent.logging.AgentLogger;
+import org.hotswap.agent.plugin.spring.core.*;
+import org.hotswap.agent.plugin.spring.files.PropertyReload;
+import org.hotswap.agent.plugin.spring.files.XmlBeanDefinitionScannerAgent;
+import org.hotswap.agent.plugin.spring.getbean.ProxyReplacer;
+import org.hotswap.agent.plugin.spring.utils.ResourceUtils;
+import org.hotswap.agent.util.AnnotationHelper;
+import org.hotswap.agent.util.spring.util.ClassUtils;
+import org.hotswap.agent.util.spring.util.ObjectUtils;
+import org.springframework.beans.PropertyValue;
+import org.springframework.beans.factory.NoSuchBeanDefinitionException;
+import org.springframework.beans.factory.annotation.AnnotatedGenericBeanDefinition;
+import org.springframework.beans.factory.config.*;
+import org.springframework.beans.factory.support.*;
+import org.springframework.context.annotation.AnnotationBeanNameGenerator;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.hotswap.agent.plugin.spring.utils.RegistryUtils.maybeRegistryToBeanFactory;
+import static org.hotswap.agent.util.ReflectionHelper.get;
+
+/**
+ * Reload spring beans.
+ */
+public class SpringBeanReload {
+    private static AgentLogger LOGGER = AgentLogger.getLogger(SpringBeanReload.class);
+
+    private AtomicBoolean isReloading = new AtomicBoolean(false);
+
+    // it is synchronized set because it is used synchronized block
+    private Set<Class<?>> classes = Collections.newSetFromMap(new ConcurrentHashMap<>());
```

Comment:
make 'classes', 'properties', 'xmls', 'newScanBeanDefinitions' final, since you put 'synchronized' against them.